--- openvswitch-2.3.0/lib/dh_set0_pqg.c	1969-12-31 19:00:00.000000000 -0500
+++ openvswitch-2.3.0/lib/dh_set0_pqg.c	2021-10-23 07:41:39.973218952 -0400
@@ -0,0 +1,24 @@
+#ifndef HAVE_DH_SET0_PQG
+int
+DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
+{
+        if ((dh->p == NULL && p == NULL) || (dh->g == NULL && g == NULL))
+                return 0;
+
+        if (p != NULL) {
+                BN_free(dh->p);
+                dh->p = p;
+        }
+        if (q != NULL) {
+                BN_free(dh->q);
+                dh->q = q;
+        }
+        if (g != NULL) {
+                BN_free(dh->g);
+                dh->g = g;
+        }
+
+        return 1;
+}
+#endif /* HAVE_DH_SET0_PQG */
+
