diff --git a/third_party/rapidjson/include/rapidjson/document.h b/third_party/rapidjson/include/rapidjson/document.h
--- a/third_party/rapidjson/include/rapidjson/document.h
+++ b/third_party/rapidjson/include/rapidjson/document.h
@@ -1936,7 +1936,7 @@
         if (count) {
             GenericValue* e = static_cast<GenericValue*>(allocator.Malloc(count * sizeof(GenericValue)));
             SetElementsPointer(e);
-            std::memcpy(e, values, count * sizeof(GenericValue));
+            std::memcpy((void*)e, (const void*)values, count * sizeof(GenericValue));
         }
         else
             SetElementsPointer(0);
@@ -1949,7 +1949,7 @@
         if (count) {
             Member* m = static_cast<Member*>(allocator.Malloc(count * sizeof(Member)));
             SetMembersPointer(m);
-            std::memcpy(m, members, count * sizeof(Member));
+            std::memcpy((void*)m, (const void*)members, count * sizeof(Member));
         }
         else
             SetMembersPointer(0);