diff --git a/third_party/tbb-2018_U5/build/linux.gcc.inc.orig b/third_party/tbb-2018_U5/build/linux.gcc.inc
index 786c415..96e3bf1 100644
--- a/third_party/tbb-2018_U5/build/linux.gcc.inc.orig
+++ b/third_party/tbb-2018_U5/build/linux.gcc.inc
@@ -56,13 +56,13 @@ endif
 
 # gcc 5.0 and later have -Wsuggest-override option
 # enable it via a pre-included header in order to limit to C++11 and above
-ifneq (,$(shell gcc -dumpversion | egrep  "^([5-9])"))
+ifneq (,$(shell g++ -dumpversion | egrep  "^([5-9])"))
     INCLUDE_TEST_HEADERS = -include $(tbb_root)/src/test/harness_preload.h
 endif
 
 # gcc 6.0 and later have -flifetime-dse option that controls
 # elimination of stores done outside the object lifetime
-ifneq (,$(shell gcc -dumpversion | egrep  "^([6-9])"))
+ifneq (,$(shell g++ -dumpversion | egrep  "^([6-9])"))
     # keep pre-contruction stores for zero initialization
     DSE_KEY = -flifetime-dse=1
 endif
